### ISSUE TITLE
Remote Tidy: Cannot get platform from install target if platform config is regex

### ISSUE THREAD

**Issue by wxtim** [cylc/cylc-flow#5429]
### Synopsis

Remote tidy logic gets a platform from install targets recorded in `TaskRemoteMgr.remote_init_map`.

Currently it uses the `globalconfig.get(['platforms'])` and attempts to do an exact match, so if platform is a regex and install target is not (it shouldn't be) then it's possible not to find a matching platform.

The platform name regex in the global config is not the platform name, and unless it has exactly one possible expansion we cannot infer the platform from an install target - we'd need to know the platform name as well.

### Reproducible Example

```
# global.cylc
[platforms]
  [[someremote0.]]
```

```
# chunk of flow.cylc
[runtime]
  [[task]]
    platform = someremote01
```

Will result in  `UnboundLocalError: local variable 'platform' referenced before assignment`.

### Where to look in the codebase

- `cylc.flow.task_remote_mgr.py` ~ line 300, method `remote_tidy`. https://github.com/cylc/cylc-flow/blob/8d7cfcbeba961dda951c18873fd561573ccc7e15/cylc/flow/task_remote_mgr.py#L304-L306
- `cylc.flow.platforms.py` ~ line 670, method `get_all_platforms_for_install_target`. https://github.com/cylc/cylc-flow/blob/8d7cfcbeba961dda951c18873fd561573ccc7e15/cylc/flow/platforms.py#L652

### Also Check

- Are there other instances of `globalconfig.get(['platforms'])` as we cannot be sure they are safe.
- Are there other uses of `get_all_platforms_for_install_target`

### Expected Behaviour

- Not attempting to get platform from install target.
- Only getting a platform from the install target if the platform name regex in the global config has one and only one possible match.

**Comment by MetRonnie** [comment]
I haven't understood the problem here. What is wrong with using a platform configuration that has the right install target, if it is just using a regex pattern instead of exact platform name?

**Comment by wxtim** [comment]
> I haven't understood the problem here. What is wrong with using a platform configuration that has the right install target, if it is just using a regex pattern instead of exact platform name?

If we have `install target = foo42` and `[[foo\d\d]]` what is the platform we want to use? We can use the install target, which we know is a valid host, but we cannot recover the platform name with this info.

**Comment by oliver-sanders** [comment]
> then why not use that platform configuration? Why is a platform "name" necessary?

In the example you gave, there are no `hosts` configured so `hosts` are defaulted from the platform name which is a regex.

If the `hosts` are not defined, then we can't reasonably use this platform because we would have to run through every possible match of the pattern to find the hosts.

If the platform-name is defined as a regex, then it's a bit messy to use the platform because we don't actually know what it's called for logging etc.

**Comment by oliver-sanders** [comment]
Closed by #5445
